src/transports/smart.c

```diff
@@ -66,6 +66,80 @@ static int git_smart__set_callbacks(
 	return 0;
 }
 
+int http_header_name_length(const char *http_header)
+{
+	const char *colon = strchr(http_header, ':');
+	if (!colon)
+		return 0;
+	return colon - http_header;
+}
+
+bool is_malformed_http_header(const char *http_header)
+{
+	const char *c;
+	int name_len;
+
+	// Disallow \r and \n
+	c = strchr(http_header, '\r');
+	if (c)
+		return true;
+	c = strchr(http_header, '\n');
+	if (c)
+		return true;
+
+	// Require a header name followed by :
+	name_len = http_header_name_length(http_header);
+	if (name_len < 1)
+		return true;
+
+	return false;
+}
+
+static char *forbidden_custom_headers[] = {
+	"User-Agent",
+	"Host",
+	"Accept",
+	"Content-Type",
+	"Transfer-Encoding",
+	"Content-Length",
+};
+
+bool is_forbidden_custom_header(const char *custom_header)
+{
+	unsigned long i;
+	int name_len = http_header_name_length(custom_header);
+
+	// Disallow headers that we set
+	for (i = 0; i < ARRAY_SIZE(forbidden_custom_headers); i++)
+		if (strncmp(forbidden_custom_headers[i], custom_header, name_len) == 0)
+			return true;
+
+	return false;
+}
+
+static int git_smart__set_custom_headers(
+	git_transport *transport,
+	const git_strarray *custom_headers)
+{
+	transport_smart *t = (transport_smart *)transport;
+	size_t i;
+
+	for (i = 0; i < custom_headers->count; i++) {
+		if (is_malformed_http_header(custom_headers->strings[i])) {
+			giterr_set(GITERR_INVALID, "custom HTTP header '%s' is malformed", custom_headers->strings[i]);
+			return -1;
+		}
+		if (is_forbidden_custom_header(custom_headers->strings[i])) {
+			giterr_set(GITERR_INVALID, "custom HTTP header '%s' is already set by libgit2", custom_headers->strings[i]);
+			return -1;
+		}
+	}
+
+	t->custom_headers = custom_headers;
```

Comment:
This assumes that the `git_strarray *custom_headers` won't go away before the transport uses it. Is that valid? If you do this:

```c
git_remote *remote;
do_fetch_with_headers(remote);
do_fetch(remote);

void do_fetch_with_headers(git_remote *remote) {
  git_fetch_options opts = /* ... */;
  opts.custom_headers = /* ... */;
  git_remote_fetch(remote, _, &opts, _);
}

void do_fetch(git_remote *remote) {
  git_remote_fetch(remote, _, NULL, _);
}
```

Does the second `git_remote_fetch` reuse any of the transport stuff from the first one? If it does reuse the transport, would it be better to do a `git_strarray_copy` instead of just copying the pointer? I had a hard time figuring this out by reading the code. I can take another look at it, but I'd appreciate another set of :eyes:.